core/src/test/java/org/apache/iceberg/data/avro/TestDecoderResolver.java

```diff
@@ -0,0 +1,167 @@
+
+    fileSchema1 = null;
+    checkCachedSize(1);
+
+    fileSchema2 = null;
+    checkCachedSize(0);
+  }
+
+  @Test
+  public void testDecoderCachingReadSchemaNotSameAsFileSchema() throws Exception {
+    Decoder dummyDecoder = DecoderFactory.get().binaryDecoder(new byte[] {}, null);
+    Schema fileSchema = avroSchema();
+    Schema readSchema = avroSchema("manifest_path", "manifest_length", "partition_spec_id");
+    ResolvingDecoder resolvingDecoder =
+        DecoderResolver.resolve(dummyDecoder, readSchema, fileSchema);
+
+    assertThat(DecoderResolver.DECODER_CACHES.get().size()).isEqualTo(1);
+    assertThat(DecoderResolver.DECODER_CACHES.get().get(readSchema).size()).isEqualTo(1);
+    checkCached(readSchema, fileSchema);
+
+    // Equal but new one
+    Schema fileSchema1 = avroSchema();
+    Schema readSchema1 = avroSchema("manifest_path", "manifest_length", "partition_spec_id");
+    assertThat(fileSchema1).isEqualTo(fileSchema);
+    assertThat(readSchema1).isEqualTo(readSchema);
+    ResolvingDecoder resolvingDecoder1 =
+        DecoderResolver.resolve(dummyDecoder, readSchema1, fileSchema1);
+    assertThat(resolvingDecoder1).isNotSameAs(resolvingDecoder);
+
+    assertThat(DecoderResolver.DECODER_CACHES.get().size()).isEqualTo(2);
+    assertThat(DecoderResolver.DECODER_CACHES.get().get(readSchema1).size()).isEqualTo(1);
+    checkCached(readSchema1, fileSchema1);
+
+    // New read schema
+    Schema readSchema2 = avroSchema("manifest_path", "manifest_length");
+    ResolvingDecoder resolvingDecoder2 =
+        DecoderResolver.resolve(dummyDecoder, readSchema2, fileSchema);
+    assertThat(resolvingDecoder2).isNotSameAs(resolvingDecoder);
+
+    assertThat(DecoderResolver.DECODER_CACHES.get().size()).isEqualTo(3);
+    assertThat(DecoderResolver.DECODER_CACHES.get().get(readSchema2).size()).isEqualTo(1);
+    checkCached(readSchema2, fileSchema);
+
+    checkCachedSize(3);
+
+    readSchema = null;
+    checkCachedSize(2);
+
+    readSchema1 = null;
+    checkCachedSize(1);
+
+    readSchema2 = null;
+    checkCachedSize(0);
+  }
+
+  private Schema avroSchema(String... columns) {
+    if (columns.length == 0) {
+      return AvroSchemaUtil.convert(ManifestFile.schema(), "manifest_file");
+    } else {
+      return AvroSchemaUtil.convert(ManifestFile.schema().select(columns), "manifest_file");
+    }
+  }
+
+  private void checkCached(Schema readSchema, Schema fileSchema) {
+    assertThat(DecoderResolver.DECODER_CACHES.get().containsKey(readSchema)).isTrue();
+    assertThat(DecoderResolver.DECODER_CACHES.get().get(readSchema).containsKey(fileSchema))
+        .isTrue();
+  }
+
+  private int getActualSize() {
+    // The size of keys included the GCed keys
+    Set<Schema> keys = DecoderResolver.DECODER_CACHES.get().keySet();
+    Set<Schema> identityKeys = Sets.newIdentityHashSet();
+    identityKeys.addAll(keys);
+    return identityKeys.size();
+  }
+
+  private void checkCachedSize(int expected) {
```

Comment:
wondering if this can be flaky on CI builds where servers are quite often overloaded. we can keep an eye on this. not a concern right now.